kliff/models/parameter.py

```diff
@@ -1,705 +1,430 @@
+    def get_numpy_array_model_space(self) -> np.ndarray:
+        """Get a numpy array of parameters in the model space.
 
-        Args:
-            index: index of the component
-            v: upper bound value
-        """
-        self._upper_bound[index] = v
+        This method should be uses for getting the numpy array of parameters where the
+        ``Parameters`` class might not work, for feeding values to the model.
 
-    @property
-    def name(self) -> str:
-        """
-        Name of the parameter.
-        """
-        return self._name
-
-    @property
-    def index(self) -> int:
-        """
-        Integer index of the parameter.
+        Returns:
+            A numpy array of parameters in the original space.
         """
-        return self._index
-
-    def __len__(self):
-        return len(self.value)
-
-    def __getitem__(self, i: int):
-        return self._value[i]
-
-    def __setitem__(self, i: int, v: float):
-        self._value[i] = float(v)
-
-    def as_dict(self):
-        return {
-            "@module": self.__class__.__module__,
-            "@class": self.__class__.__name__,
-            "value": np.asarray(self._value).tolist(),
-            "fixed": self._fixed,
-            "lower_bound": self._lower_bound,
-            "upper_bound": self._upper_bound,
-            "name": self._name,
-            "index": self._index,
-        }
-
-    @classmethod
-    def from_dict(cls, d):
-        return cls(
-            value=d["value"],
-            fixed=d["fixed"],
-            lower_bound=d["lower_bound"],
-            upper_bound=d["upper_bound"],
-            name=d["name"],
-            index=d["index"],
-        )
-
-
-class OptimizingParameters(MSONable):
-    """
-    A collection of parameters that will be optimized.
-
-    This can be all the parameters of a model or a subset of the parameters of a model.
-    The behavior of individual component of a parameter can also be controlled. For
-    example, keep the 2nd component of a parameters fixed, while optimizing the other
-    components.
-
-    It interacts with optimizer to provide initial guesses of parameter values; it also
-    receives updated parameters from the optimizer and update model parameters.
-
-    Args:
-        model_params: {name, parameter} all the parameters of a model. The attributes
-            of these parameters will be modified to reflect whether it is optimized.
-    """
+        if (self.transform_function is not None) and self._is_transformed:
+            return self.transform_function.inverse_transform(self)
+        else:
+            return np.asarray(self)
 
-    def __init__(self, model_params: Dict[str, Parameter]):
-        self.model_params = model_params
+    def get_numpy_array_param_space(self):
+        """Applies the transform to the parameter, and returns the transformed array."""
+        self.transform()
+        return np.asarray(self)
```

Comment:
I don't think `np.asarray(self)` will take any effect. If you want to make a copy, do `self.copy()` or `np.array(self)`?

Reply:
Replaced with np.array.
Idea was to copy and provide a pure numpy array so that it does not give any issue downstream, where some function might perform exception checks for numpy arrays.